lib/Dialect/ESI/runtime/python/esiaccel/cosim/simulator.py

```diff
@@ -116,23 +172,28 @@ def compile_commands(self) -> List[List[str]]:
   def compile(self) -> int:
     cmds = self.compile_commands()
     self.run_dir.mkdir(parents=True, exist_ok=True)
-    with (self.run_dir / "compile_stdout.log").open("w") as stdout, (
-        self.run_dir / "compile_stderr.log").open("w") as stderr:
-      for cmd in cmds:
-        stderr.write(" ".join(cmd) + "\n")
-        cp = subprocess.run(cmd,
-                            env=Simulator.get_env(),
-                            capture_output=True,
-                            text=True)
-        stdout.write(cp.stdout)
-        stderr.write(cp.stderr)
-        if cp.returncode != 0:
-          print("====== Compilation failure:")
-          if self.UsesStderr:
-            print(cp.stderr)
-          else:
-            print(cp.stdout)
-          return cp.returncode
+    for cmd in cmds:
+      ret = self._start_process_with_callbacks(
+          cmd,
+          env=Simulator.get_env(),
+          cwd=None,
+          stdout_cb=self._compile_stdout_cb,
+          stderr_cb=self._compile_stderr_cb,
+          wait=True)
+      if isinstance(ret, int) and ret != 0:
+        print("====== Compilation failure")
+
+        # If we have the default file loggers, print the compilation logs to
+        # console. Else, assume that the user has already captured them.
+        if self.UsesStderr:
+          if hasattr(self, "_compile_stderr_log"):
+            self._compile_stderr_log.seek(0)
+            print(self._compile_stderr_log.read())

```

Comment:
As long as we're reading from the file, is there a way to stream the file directly to stdout (instead of reading the whole file into memory and then printing it)?

Reply:
As in, stream it and print it during compilation? If that's the behavior we want, that could just be added to the logging callback for when the user didn't provide a logging callback.

Reply:
No, I meant this line and the one above it. I'm assuming that `read()` reads the whole file into memory, though it may be more intelligent than that.

But this is very minor.